### PR TITLE
Graph Compiler Framework

### PR DESCRIPTION
**Pull request by Kristoff-starling**
a draft PR tracking the progress of graph compiler implementatoin

### CODE REVIEW

**Comment by Kristoff-starling** [comment]
@Romero027 Could you please take a look at the PR? Thanks!

**Review by Romero027 — COMMENTED** [review]
Thanks @Kristoff-starling ! Can you provide instructions on how to run the compiler inthe README file?

**Comment by Kristoff-starling** [comment]
I've updated usage instructions in `compiler/graph/README.md`. However, it's a bit tricky to correctly fire up the entire hotel reservation application for now, which makes it hard to fully exercise the compiler.
A workaround could be temporarily removing all the functions interacting with remote containers in `compiler/graph/backend/utils.py` (but keeping the interfaces, for functions having return value, simply return "0"). In this way, the compiler will not send commands to remote containers while still generating all the scripts inside `graph/gen`.

**Comment by Kristoff-starling** [comment]
@Romero027 Now you can use `python3 main.py --pseudo_element --no_run_container` to run the compiler without setting up the hotel reservation application. (But you still need to clone the phoenix repository and install necessary dependencies by following the instructions in `compiler/graph/README.md` :)

File: compiler/graph/README.md [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+```bash
+# in compiler/
+python3 main.py [--verbose] [--pseudo_element] [--spec path_to_spec] [--backend BACKEND] [--no_run_container]
+```
+* `--verbose`: makes the compiler more chatty.
+* `--pseudo_element`: use the pseudo element compiler provided by the graph compiler, which reads element properties in `element/property/` and copy existing implementations from the phoenix local repository.
+* `--spec path_to_spec`: if not specified, `example_spec/dummy.yml` will be used by default.
+* `--backend BACKEND`: if not specified, the graph compiler generates mrpc scripts by default.
+* `--no_run_container`: if used, the graph compiler will not send remote commands into the container but only print the commands on the screen.
```
Comment (Romero027) [inline compiler/graph/README.md:48]: I think `--dry-run` might be a better name

File: compiler/graph/backend/mrpc.py [hunk 2]
```diff
@@ -0,0 +1,271 @@
+from __future__ import annotations
```
Comment (Romero027) [inline compiler/graph/backend/mrpc.py:1]: can you add 1-2 paragraphs describing what this file does?

**Review by Romero027 — APPROVED** [review]
Thanks @Kristoff-starling. This is a very good start. One small request is to add more comments as you work on this (at least a few paragraphs in each file describing what it does). 

Feel free to merge.

File: compiler/graph/backend/mrpc.py (hunk 2 above)
Comment (Kristoff-starling) [reply compiler/graph/backend/mrpc.py:5]: I've added descriptions and docstrings in files/functions/classes, and it should be easier to understand and reuse :)

**Pull request merged**